ISPN-7114: Consistency Checker and Automatic Merge Policies

ryanemerson opened the pull request:
@tristantarrant @danberindei @rvansa I have implemented a basic version of the ConflictResolutionManager which allows for a lazy stream of conflicts to be consumed and for all versions of a given key to be returned. Merge resolution hasn't been implemented yet. Could you please have a quick look at my approach and make sure I'm not barking up the wrong tree :)

Tests to follow as well.

rvansa commented on core/src/main/java/org/infinispan/conflict/resolution/StateReceiverImpl.java:
```diff
@@ -0,0 +1,134 @@
+   private RpcManager rpcManager;
+   private long transferTimeout;
+
+   @GuardedBy("keyReplicaMapLock")
```
there's no such field

rvansa commented on core/src/main/java/org/infinispan/conflict/resolution/StateReceiverImpl.java:
```diff
@@ -0,0 +1,134 @@
+
+   @Override
+   public CompletableFuture<List<Map<Address, CacheEntry<K, V>>>> getAllReplicasForSegment(int segmentId) {
+      Set<Integer> segments = new HashSet<>();
```
Collections.singleton

rvansa commented on core/src/main/java/org/infinispan/conflict/resolution/StateReceiverImpl.java:
```diff
@@ -0,0 +1,134 @@
+   }
+
+   private void addKeyToReplicaMap(Address address, CacheEntry<K, V> ice) {
+      Map<Address, CacheEntry<K, V>> keyMap = keyReplicaMap.getOrDefault(ice.getKey(), new HashMap<>());
```
`keyReplicaMap.computeIfAbsent(ice.getKey(), k -> new HashMap<>()).put(address, ice);`

rvansa commented on core/src/main/java/org/infinispan/security/impl/SecureCacheImpl.java:
```diff
@@ -698,6 +699,12 @@ public void removeGroup(String groupName) {
 
    @Override
+   public ConflictResolutionManager<K, V> getConflictResolutionManager() {
+      authzManager.checkPermission(AuthorizationPermission.ADMIN);
```
Isn't this more like `ALL_READ + ALL_WRITE`?

rvansa commented on core/src/main/java/org/infinispan/conflict/resolution/StateReceiverImpl.java:
```diff
@@ -0,0 +1,134 @@
+         InboundTransferTask transferTask = transferTaskMap.get(sender);
+         boolean lastChunkReceived = false;
+         for (StateChunk chunk : stateChunks) {
+            transferTask.onStateReceived(chunk.getSegmentId(), chunk.isLastChunk());
```
Can't this trigger any @DataRehashed listeners?

ryanemerson replied:
@rvansa I don't believe so as `CacheNotifier::notifyDataRehashed` is only called in  `StateConsumerImpl`.

rvansa commented on core/src/main/java/org/infinispan/statetransfer/StateRequestCommand.java:
```diff
@@ -133,12 +140,21 @@ public byte getCommandId() {
       switch (type) {
-         case GET_TRANSACTIONS:
          case START_STATE_TRANSFER:
+            output.writeBoolean(applyState);
```
I would rather add new type than a flag.

rvansa commented on core/src/main/java/org/infinispan/conflict/resolution/StateReceiverImpl.java:
```diff
@@ -0,0 +1,134 @@
+
+      CacheTopology cacheTopology = localTopologyManager.getCacheTopology(cache.getName());
+      ConsistentHash hash = cacheTopology.getWriteConsistentHash();
+      List<Address> replicas = hash.locateOwnersForSegment(segmentId);
```
This does not deal well with concurrent topology changes (you should at least throw exception when something like that happens). This way you could contact owners that are already losing the ownership and miss new ones. I am not sure in what way will the InboundTransferTask fail if there's a topology change on the owner, but it should be handled somehow.

rvansa commented on core/src/main/java/org/infinispan/conflict/resolution/StateReceiverImpl.java:
```diff
@@ -0,0 +1,134 @@
+         }
+
+         if (lastChunkReceived) {
+            transferFuture.complete(null);
```
It seems that this is the only use of transferFuture - could you call `getAndClearKeyReplicaMap` directly?

rvansa commented on core/src/main/java/org/infinispan/conflict/resolution/StateReceiverImpl.java:
```diff
@@ -0,0 +1,134 @@
+   }
+
+   @Override
+   public CompletableFuture<List<Map<Address, CacheEntry<K, V>>>> getAllReplicasForSegment(int segmentId) {
```
It seems that if you call this method concurrently (or whenever before the future completes), transfer tasks from one call will overwrite the other.

ryanemerson replied:
@rvansa I agree. How do you think we should handle such a case? As statereceiver is created on a per cache basis and is only invoked via DefaultConflictResolutionManager::getConflicts(), I think it might be overkill to allow for concurrent requests on this method. Furthermore, I'm not sure allowing concurrent executions of `getConflicts` on the same node should be allowed as this could cause problems when it comes to conflict resolution.

Is it sufficient to simply throw an IllegalStateException in the event that concurrent executions of getConflict (and also getAllReplicasForSegment) occur?

ryanemerson commented:
@rvansa Thanks for your comments. I have updated the PR based upon your suggestions and added some tests. Sorry about the slow response, I was working on other things the last few weeks.

galderz commented:
Needs rebasing. I had forgotten to click button to submit review (it's on there now).

This one should be quick to introduce once sorted out, so leaving it open.

galderz commented:
@ryanemerson Why is marked as preview? Is there more work that needs to be done here?

galderz commented:
Ah yes, are the tests that are supposed to be added in? Closing for now, complete with tests and reopen then.

ryanemerson commented:
@galderz It was marked as a preview, as the consistency check was only part of the work. The plan is to also implement automatic merge of conflicts upon partition merges (see [here).](https://github.com/infinispan/infinispan/wiki/Conflict-resolution) But, now I'm thinking about it maybe it makes sense to just get the consistency checker in first and issue a separate PR for the other stuff so this doesn't get too bloated. @galderz, @rvansa WDYT?

Regarding tests, there is `ConflictResolutionManagerTest` and `StateReceiverTest`.

rvansa commented on core/src/test/java/org/infinispan/conflict/resolution/ConflictResolutionManagerTest.java:
```diff
@@ -0,0 +1,133 @@
+      }
+   }
+
+   private void introduceCacheConflicts() {
```
I would appreciate some entries completely missing on given node, too.

ryanemerson replied:
Done

rvansa commented on core/src/main/java/org/infinispan/conflict/resolution/EntryMergePolicy.java:
```diff
@@ -0,0 +1,14 @@
+ * @author Ryan Emerson
+ */
+@Experimental
+public interface EntryMergePolicy<K, V> {
```
This class is not used anywhere (yet). If you want to have "final" API without implementing the logic, you should add this to configuration, too. Or just omit the methods from `ConflictResolutionManager` for the time being.

ryanemerson replied:
I have removed EntryMergePolicy for now and I will re-introduce when required.

rvansa commented on core/src/main/java/org/infinispan/conflict/resolution/DefaultConflictResolutionManager.java:
```diff
@@ -0,0 +1,159 @@
+         Thread.currentThread().interrupt();
+         throw new CacheException(e);
+      } catch (ExecutionException e) {
+         throw new CacheException(e.getCause());
```
Actually if you use `SKIP_OWNERSHIP_CHECK`, it won't throw OTE (if I am not mistaken).

ryanemerson commented on core/src/main/java/org/infinispan/conflict/resolution/StateReceiverImpl.java:
```diff
@@ -0,0 +1,230 @@
+         }
+
+         this.segmentId = segmentId;
+         ConsistentHash hash = cacheTopology.getReadConsistentHash();
```
Ok, I've updated the code so that `getWriteConsistentHash()` is always used as it is more explicit than `getConsistentHash()`

ryanemerson commented:
@rvansa I have made the requested changes. If a user attempts to initiate `getConflicts` or `getAllVersions(key)` when ST is in progress, than an `IllegalStateException` is thrown. Furthermore, I have added the `isResolutionPossible()` methods for user's to determine if it is acceptable for them to proceed.

rvansa commented on core/src/main/java/org/infinispan/conflict/resolution/DefaultConflictResolutionManager.java:
```diff
@@ -0,0 +1,177 @@
+
+   @Override
+   public Map<Address, InternalCacheValue<V>> getAllVersions(K key) {
+      if (stateTransferManager.isStateTransferInProgressForKey(key))
```
IIUC, this returns true when ST is in progress AND the key is being transferred to current node. If this node won't be an owner, it will return false.

rvansa replied:
And since you're not checking for topology changes, the ST could start right after this check and the entries could be moved around during the operation (with cache topology changing).

rvansa commented on core/src/main/java/org/infinispan/conflict/resolution/DefaultConflictResolutionManager.java:
```diff
@@ -0,0 +1,177 @@
+      if (stateTransferManager.isStateTransferInProgressForKey(key))
+         throw log.getConflictStateTransferInProgress(key);
+
+      List<Address> keyReplicaNodes = distributionManager.getWriteConsistentHash().locateOwners(key);
```
FYI (doesn't need a change): if you decide to ban operation during ST, writeCH will be identical to readCH. Using writeCH has the advantage of being able to find all owners, including those we can't read from.

rvansa commented on core/src/test/java/org/infinispan/conflict/resolution/ConflictResolutionManagerTest.java:
```diff
@@ -0,0 +1,188 @@
+      TestingUtil.replaceComponent(cache0, RpcManager.class, controlledRpcManager, true);
+      controlledRpcManager.blockAfter(StateRequestCommand.class);
+
+      StateConsumer sc = TestingUtil.extractComponent(cache0, StateConsumer.class);
```
This does not look like really starting state transfer, you just change topology, without real rehash happening.

rvansa commented on core/src/test/java/org/infinispan/conflict/resolution/ConflictResolutionManagerTest.java:
```diff
@@ -0,0 +1,188 @@
+      StateConsumer sc = TestingUtil.extractComponent(cache0, StateConsumer.class);
+
+      sc = spy(sc);
+      doReturn(true).when(sc).isStateTransferInProgressForKey(any());
```
IMO too synthetic test, you just check that a method is called, but it's not returning the actual value.

ryanemerson commented:
@rvansa I have updated `ConflictResolutionManager.getAllVersions(...)` so that it fails if ST is in progress at the start of the method, or if a rebalance occurs during its invocation. If concurrent invocations of this method occur on a given key, then the same result is returned for both invocations. Of course, consecutive invocations of the method will start a new attempt to retrieve all versions of a key's entry. 

I have also updated the tests based upon your suggestions as well as adding a few additional tests to cover the changes stated above.

galderz commented:
@rvansa Are you happy with the tests added by @ryanemerson ?

rvansa commented on core/src/test/java/org/infinispan/conflict/resolution/ConflictResolutionManagerTest.java:
```diff
@@ -0,0 +1,265 @@
+package org.infinispan.conflict.resolution;
+
+import static org.infinispan.test.TestingUtil.wrapPerCacheInboundInvocationHandler;
+import static org.testng.Assert.assertEquals;
```
I think that AssertJUnit should be preferred.

rvansa commented on core/src/test/java/org/infinispan/conflict/resolution/ConflictResolutionManagerTest.java:
```diff
@@ -0,0 +1,265 @@
+   public void testGetAllVersionsDuringStateTransfer() {
+      createAndSplitCluster();
+      blockStateTransferCompletion();
+      partition(0).merge(partition(1));
```
Looking into `Partition.merge`, it seems that the code is waiting until the rebalance/state transfer completes. How can that complete, when you've blocked the state transfer? Can you be sure that the ST has already started?

rvansa commented:
@galderz Thanks for reminding me of this PR, I left it waiting for a while. I have still few comments.

tristantarrant commented on core/src/main/java/org/infinispan/AdvancedCache.java:
```diff
@@ -556,4 +557,10 @@
+    * Get the {@link ConflictResolutionManager} associated with this cache.
+    * @return the conflict manager for this cache.
+    */
+   ConflictResolutionManager<K, V> getConflictResolutionManager();
```
Indeed, the CounterManager approach is the one to follow.

ryanemerson commented:
@rvansa Thanks for the feedback. I have updated and made the following changes:

- Changed from `ConflictResolutionManager` -> `ConflictManager`
- Split packages into public and private (`impl`)
- No longer expose the `ConflictManager` via the `AdvancedCache` interface
- Removed `onTopologyUpdate` methods and instead utilise the `@TopologyChanged` and `@DataRehashed` listeners
- Updated `ConflictManager.getAllVersions` so that it handles topology changes/rehashes without throwing an exception. See `ConflictManager` interface for details.

ryanemerson commented:
@rvansa I'll add the documentation to this PR. I have to make a small change to `DefaultCacheManager::start` aswell as it stores RPCOptions as a constant. My plan is to follow a similar approach to https://github.com/infinispan/infinispan/pull/4897

ryanemerson commented:
@rvansa @danberindei I have implemented conflict resolution as detailed [in the design doc](https://github.com/infinispan/infinispan-designs/blob/master/Conflict-resolution.md). The initial commit is very similar to the one origianlly approved by @rvansa, however there are a few minor changes to the original commit which were necessary due to rebasing, version number change etc. 

This is very much a preview, as I plan to add additional tests for automatic merging (more scenarios) as well as implementing parallel segment requests in `DefaultConflictManager::ReplicaSpliterator` and `StateReceiver::getAllReplicasForSegment`

galderz commented:
Needs rebasing

slaskawi commented:
@ryanemerson Is this ready to be integrated? Looks fine...

ryanemerson commented:
@slaskawi As much as I would love that to be the case, definitely not. @rvansa reviewed most of the code exposed via `ConflictManager` to the user, however since his last review I have had to rebase to take into account https://github.com/infinispan/infinispan/pull/4862 as well as adding automatic conflict resolution on merge policies which is a lot of work in itself.

rvansa commented:
@ryanemerson So are you still working on that (expecting major changes), or should I do a review?

ryanemerson commented:
@rvansa A review would be great! I don't plan on any major changes until you or @danberindei tell me otherwise :stuck_out_tongue_winking_eye:

ryanemerson commented:
Thanks for the feedback @rvansa. The merge policies are applied in [DefaultConflictManager.java](https://github.com/ryanemerson/infinispan/blob/1ebd3d01bdce54bb23fd448d9f55d954a4b907a7/core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java)

ryanemerson commented on core/src/main/java/org/infinispan/configuration/cache/PartitionHandlingConfigurationBuilder.java:
```diff
@@ -27,8 +26,19 @@ public PartitionHandlingConfigurationBuilder(ClusteringConfigurationBuilder buil
+      return this;
+   }
+
+   public PartitionHandlingConfigurationBuilder type(PartitionHandling partitionHandling) {
```
@rvansa I agree that the naming can be improved. I will go with `whenSplit(DENY_READ_WRITES | ALLOW_READS | ALLOW_ALL)`

rvansa replied:
maybe `ALLOW_READ_WRITE`(`S`), to match `DENY_READ_WRITE`, wdyt?

ryanemerson replied:
Sure @rvansa that makes sense

rvansa commented on core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java:
```diff
@@ -0,0 +1,504 @@
+         if (trace) log.tracef("Putting entry for key %s at address %s with topology id %s", entry, address, topologyId);
+      }
+
+      Map<Address, Response> responseMap = rpcManager.invokeRemotely(cmdMap, rpcOptions);
```
It seems that all you do for an unsuccessful update is logging the failure (that's fine). Could you do the invocation asynchronously, then? Because this way the conflict resolution needs to wait for each roundtrip.

rvansa commented on core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java:
```diff
@@ -0,0 +1,504 @@
+      Map<Address, ReplicableCommand> cmdMap = new HashMap<>();
+      for (Address address : keyOwners) {
+         int topologyId = mergeInfo.getTopologyId(address);
+         PutKeyValueCommand cmd = commandsFactory.buildPutKeyValueCommand(entry.getKey(), entry.getValue(), entry.getMetadata(), remoteFlags);
```
`remoteFlags` does not contain `CACHE_MODE_LOCAL`, so the primary owner will forward that to all backups as well, doesn't it?
I am rather unsure about the `SKIP_OWNERSHIP_CHECK` - backup owners don't acquire locks, and executing this concurrently to another update is dangerous. Is there anything that would prevent further updates, or is it just the app's responsibility (is this documented?). Btw., the forwarded write from primary is a concurrent update, too.

ryanemerson commented:
@rvansa I have introduced `CacheTopology.Phase.CONFLICT_RESOLUTION` which occurs before `READ_OLD_WRITE_ALL`. This phase sets the current CH to the one set by the `maxTopology` that is encountered during a partition merge and sets the pending/union CH to the ConsistentHash that will be utilised after the rebalance. As a result of applying this topology, Removes/Puts during Conflict resolution now utilise the normal Cache interface methods, with locking etc applied via the TriangleInterceptor.

The latest commit enables segments to be retrieved in parallel by the ConflictManager, however the stream is currently set to be executed sequentially as I still need to implement a way for the StateTransferThreadPool threads to execute the parallel stream.

rvansa commented:
@ryanemerson I guess that Dan will be happy with another phase :) Btw., `currentCH`/`pendingCH`/`unionCH` are topologies from the POV of ST, I think you meant `readCH`/`writeCH` - these are the interfaces for interceptors etc. I think that during conflict resolution nodes should have `readCH` same as during partition but `writeCH` should be `union(oldWriteCH, pendingCH)` (where `pendingCH` is the CH after rebalance).
I'll check out the actual code.

rvansa commented on core/src/main/java/org/infinispan/conflict/impl/StateReceiverImpl.java:
```diff
@@ -138,34 +153,32 @@ void onDataRehash(DataRehashedEvent dataRehashedEvent) {
             }
          }
-         segmentRequestFuture = CompletableFuture
+         future = CompletableFuture
```
I would also think about limiting the degree of parallelism here. I recall issues in 32 node tests, where a node joining the cluster requested state from all other nodes at once - and the other 31 nodes started sending it, effectively DDoSing the poor joiner.

rvansa commented on core/src/main/java/org/infinispan/topology/ClusterTopologyManagerImpl.java:
```diff
@@ -422,8 +422,8 @@ private ClusterCacheStatus initCacheStatusIfAbsent(String cacheName) {
          Configuration cacheConfiguration = cacheManager.getCacheConfiguration(cacheName);
-         boolean resolveConflictsOnMerge = cacheConfiguration.clustering().partitionHandling().mergePolicy() != null;
          PartitionHandling  partitionHandling = cacheConfiguration != null ? cacheConfiguration.clustering().partitionHandling().whenSplit() : null;
+         boolean resolveConflictsOnMerge = partitionHandling == PartitionHandling.ALLOW_READ_WRITES && cacheConfiguration.clustering().partitionHandling().mergePolicy() != null;
```
If there's no default merge policy, we should throw an exception in the `ConfigurationBuilder.validate()` when the partition handling is set to ARW.

ryanemerson replied:
I did think about that, however I thought it might be advantageous for us to be able to allow users to keep the existing merge behaviour, in the event a user depends on the old semantics, by simply specifying a null policy. That said I'm happy to remove this option if you think this is unnecessary.

rvansa commented on core/src/main/java/org/infinispan/configuration/parsing/Parser.java:
```diff
@@ -2575,7 +2575,8 @@ public static Properties parseProperties(final XMLExtendedStreamReader reader) t
       PREFERRED_ALWAYS(MergePolicies.PREFERRED_ALWAYS),
-      PREFERRED_NON_NULL(MergePolicies.PREFERRED_NON_NULL);
+      PREFERRED_NON_NULL(MergePolicies.PREFERRED_NON_NULL),
+      REMOVE_ALL(MergePolicies.REMOVE_ALL);
```
Is it there just for completeness or do you have an use case?

rvansa commented on core/src/main/java/org/infinispan/partitionhandling/impl/PreferAvailabilityStrategy.java:
```diff
@@ -27,16 +26,26 @@
+      int topologyId = ((Integer)t1.getTopologyId()).compareTo(t2.getTopologyId());
+      if (topologyId != 0)
+         return topologyId;
+      return t1.getCurrentCH().getRoutingTableAsString().compareTo(t2.getCurrentCH().getRoutingTableAsString());
```
This left my eyes bleeding!

rvansa commented on core/src/main/java/org/infinispan/topology/CacheTopology.java:
```diff
@@ -127,6 +127,10 @@ else if (currentCH != null)
     */
    public ConsistentHash getReadConsistentHash() {
       switch (phase) {
+         case CONFLICT_RESOLUTION:
```
Just move the label before READ_OLD_WRITE_ALL, it has the same properties.

rvansa commented on core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java:
```diff
@@ -0,0 +1,459 @@
+
+      @Override
+      public Spliterator<Map<Address, InternalCacheEntry<K, V>>> trySplit() {
+         int lo = currentSegment;
```
Situation: you have 4 segments (0 - 3), now you are iterating through segment 2, so currentSegment is 3. lo is 3, mid is (3 + 4)/2 = 3 and you won't split, though you could as you haven't started going through 3 yet.

ryanemerson replied:
Good spot! Fixed with `int mid = ((lo + totalSegments) >>> 1) & ~1;

rvansa replied:
Eh, really? `3 & ~1 == 2`, and then `3 = lo  < mid = 2` is false... I think you have to test `mid < totalSegments` and use `ReplicaSpliterator(topology, mid, totalSegments)`

ryanemerson replied:
Actually I'm not sure the original situation that you describe can occur with the fix in my comment, because the original spliterator always starts with its currentSegment set to 0 and the use of `~1` ensures that the starting segment is always even. The Spliterator docs state: `If the original thread hands a spliterator off to another thread for processing, it is best if that handoff occurs before any elements are consumed with tryAdvance()`, therefore assuming the java stream implementation follows this advice, the original situation you described should never occur as the `lo` variable will always be even when `trySplit()` is called.

rvansa replied:
It's just an advice, it could ask to split it any time...  and the spliterator should split if that's possible.

rvansa replied:
So if you have 4 segments with 'infinite' elements, any sequence of `tryAdvance` and `trySplit` should end up with 4 spliterators.

rvansa commented on core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java:
```diff
@@ -0,0 +1,459 @@
+      if (trace) log.tracef("Attempting to resolve conflicts.  All Members %s, Installed topology %s, Preferred Partition %s",
+            topology.getMembers(), topology, preferredPartition);
+
+      List<CompletableFuture<V>> completableFutures =
```
Uh, won't this store a CF for each entry, not releasing anything until the end of ST? I would suggest AtomicLong (initialized with 1 and decremented once you request data for all keys) + CountDownLatch instead, as you don't care about the return values anyway. Maybe you'd need a semaphore, too, to limit the parallelism to a reasonable degree.
Another solution (more streamish) would be to call a reduce phase that would drop completed futures automatically, but it might be more tricky to tweak the `Characteristics` to make sure that you won't keep the CFs around.

rvansa replied:
Hmm, it seems that `CommitManager` will keep all keys in a `ConcurrentMap<K, ...>`, too, but this is another issue (and maybe to be expected as e.g. streams keep all keys for given segment in memory, too).

rvansa commented on core/src/main/java/org/infinispan/conflict/impl/StateReceiverImpl.java:
```diff
@@ -0,0 +1,233 @@
+
+   @Override
+   public CompletableFuture<List<Map<Address, InternalCacheEntry<K, V>>>> getAllReplicasForSegment(int segmentId, LocalizedCacheTopology topology) {
+      return requestMap.computeIfAbsent(segmentId, id -> new SegmentRequest(id, topology)).requestState();
```
I think that if this is called in parallel with `stop()`/`onDataRehash` the request may not be cancelled.

rvansa commented on core/src/main/java/org/infinispan/conflict/impl/StateReceiverImpl.java:
```diff
@@ -0,0 +1,233 @@
+               completableFutures.add(transferTask.requestSegments());
+            }
+         }
+         future = CompletableFuture
```
I would also think about limiting the degree of parallelism here. I recall issues in 32 node tests, where a node joining the cluster requested state from all other nodes at once - and the other 31 nodes started sending it, effectively DDoSing the poor joiner.

rvansa commented on core/src/main/java/org/infinispan/conflict/MergePolicies.java:
```diff
@@ -0,0 +1,20 @@
+      return (CacheEntry) otherEntries.get(0);
+   };
+
+   public static final EntryMergePolicy REMOVE_ALL = (preferredEntry, otherEntries) -> null;
```
Is it there just for completeness or do you have an use case?

ryanemerson replied:
I don't have a specific use case in mind, but as the cost of implementation is low, I figured it was a worthwhile addition. It was also necessary to ensure that entries are removed from the cache if the merge policy returns null, so this policy (or a variation of it) would have to exist in the test code.

ryanemerson commented on core/src/main/java/org/infinispan/conflict/impl/StateReceiverImpl.java:
```diff
@@ -82,45 +67,75 @@ public void start() {
+            rpcManager, commandsFactory, transferTimeout, cacheName, false);
+   }
+
+   class SegmentRequest {
```
I refactored the code to allow for concurrent segment requests, as previously the `StateReceiverImpl` only allowed a single segment request to be executing at one time.

pruivo commented on core/src/main/java/org/infinispan/partitionhandling/impl/PartitionHandlingInterceptor.java:
```diff
@@ -23,37 +19,29 @@
 
 public class PartitionHandlingInterceptor extends DDAsyncInterceptor {
    private static final Log log = LogFactory.getLog(PartitionHandlingInterceptor.class);
+   private static final boolean trace = log.isTraceEnabled();
```
unused

pruivo commented on core/src/main/java/org/infinispan/conflict/impl/StateReceiverImpl.java:
```diff
@@ -0,0 +1,233 @@
+            return;
+         }
+
+         InboundTransferTask transferTask = transferTaskMap.remove(sender);
```
remove? what if this isn't the last chunk? the next chunk wouldn't find the task

ryanemerson replied:
Good spot!

pruivo commented on core/src/test/java/org/infinispan/conflict/impl/StateReceiverTest.java:
```diff
@@ -0,0 +1,194 @@
+         cf.get();
+         fail("Expected an ExecutionExceptions caused by a CacheException");
+      } catch (ExecutionException e) {
+         assertTrue("Expected an ExecutionExceptions caused by a CacheException.", e.getCause() instanceof CacheException);
```
you could use `Exceptions.expectExecutionException();`

pruivo commented on core/src/main/java/org/infinispan/configuration/parsing/Attribute.java:
```diff
@@ -37,6 +37,7 @@
     CONFIGURATION("configuration"),
     CONSISTENT_HASH_FACTORY("consistent-hash-factory"),
     CORE_THREADS("core-threads"),
+    CUSTOM("custom"),
```
where is it used?

ryanemerson replied:
Good spot, this is a refactoring leftover

ryanemerson commented on core/src/main/java/org/infinispan/topology/ClusterTopologyManagerImpl.java:
```diff
@@ -114,7 +115,7 @@ boolean isCoordinator() {
 
 
-   private final ConcurrentMap<String, ClusterCacheStatus> cacheStatusMap = CollectionFactory.makeConcurrentMap();
+   public final ConcurrentMap<String, ClusterCacheStatus> cacheStatusMap = CollectionFactory.makeConcurrentMap();
```
Hangover from some local diagnostics, I'll fix.

ryanemerson commented on core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java:
```diff
@@ -0,0 +1,438 @@
+      resolveConflicts(installedTopology);
+   }
+
+   public void resolveConflicts(CacheTopology topology) {
```
Not when `resolveConflicts(...)` is called from the `ClusterCacheStatus`

ryanemerson commented on core/src/main/java/org/infinispan/conflict/impl/EntryMergePolicyFactory.java:
```diff
@@ -0,0 +1,22 @@
+ * @since 9.1
+ */
+@DefaultFactoryFor(classes = EntryMergePolicy.class)
+public class EntryMergePolicyFactory extends AbstractNamedCacheComponentFactory implements AutoInstantiableFactory {
```
I created this for convenience as much as anything, I'm happy to remove it and just retrieve from the configuration

rvansa replied:
+1 for removal, EntryMergePolicy is not a component

ryanemerson commented on core/src/main/java/org/infinispan/conflict/impl/NullValueEntry.java:
```diff
@@ -0,0 +1,22 @@
+ * @author Ryan Emerson
+ * @since 9.1
+ */
+class NullValueEntry extends ImmortalCacheEntry {
```
Yes. Previously I couldn't, however I have altered my approach so that it's now possible.

ryanemerson commented on core/src/main/java/org/infinispan/conflict/impl/ConflictMangerModuleLifecycle.java:
```diff
@@ -0,0 +1,27 @@
+ */
+@SuppressWarnings("unused")
+@MetaInfServices(value = ModuleLifecycle.class)
+public class ConflictMangerModuleLifecycle extends AbstractModuleLifecycle {
```
I didn't create an internal factory as the ConflictManager is only ever called internally from `ClusterCachestatus` which is created per cache inside the `ClusterTopologyManagerImpl` (this is global scope, so can't inject ConflictManager on per cache basis).

It was agreed that the ConflictManager should be exposed to the user via `ConflictManagerFactory.get(cache);`

ryanemerson replied:
Actually, I've changed approach and removed this file. Now the `ConflictManager` and `StateReceiver` are now created via the `StateTransferComponentFactory`.

danberindei replied:
+100 to make it a regular component, `ModuleLifecycle` should only be used by modules other than core.

ryanemerson commented on core/src/main/java/org/infinispan/notifications/cachelistener/CacheNotifier.java:
```diff
@@ -98,9 +98,9 @@ void notifyCacheEntryPassivated(K key, V value, boolean pre,
    void notifyTransactionRegistered(GlobalTransaction globalTransaction, boolean isOriginLocal);
 
-   void notifyDataRehashed(ConsistentHash oldCH, ConsistentHash newCH, ConsistentHash unionCH, int newTopologyId, boolean pre);
+   void notifyDataRehashed(ConsistentHash oldCH, ConsistentHash newCH, ConsistentHash unionCH, int newTopologyId, boolean pre, CacheTopology.Phase phase);
```
Good point about exposing too much to the user. Originally, before my solution introduced a new `CacheTopology.Phase,` the `StateConsumer` interacted directly with the `ConflictManager` and we decided it was best to use the Notifier. But obviously I now need to share additional information between the components. I will refactor so that the `StateConsumer` interacts directly and remove the Phase from the notifier events as before.

ryanemerson commented on core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java:
```diff
@@ -0,0 +1,438 @@
+            versionRequestMap.values().forEach(VersionRequest::cancelRequestIfOutdated);
+         }
+      } else if (!event.isPre() && event.getPhase() == CacheTopology.Phase.READ_ALL_WRITE_ALL) {
+         this.installedTopology = createLocalizedTopology(event.getConsistentHashAtEnd());
```
I've changed how this works now, so that the `StateConsumerImp` interacts directly with the `DefaultConflictManager`.

ryanemerson commented:
@pruivo @rvansa Updated based upon Pedro's feedback. Obviously the history is a mess atm, but this allows you to see the most recent changes in isolation and my intention is to squash to a single commit when everything is ready.

pruivo commented on core/src/main/java/org/infinispan/conflict/ConflictManagerFactory.java:
```diff
@@ -0,0 +1,28 @@
+         authzManager.checkPermission(AuthorizationPermission.ALL_WRITE);
+      }
+
+      return cache
```
`cache` is already an `AdvancedCache`

rvansa commented on core/src/main/java/org/infinispan/conflict/EntryMergePolicy.java:
```diff
@@ -0,0 +1,35 @@
+    * as well as when a partition merge occurs with {@link PartitionHandling#ALLOW_READ_WRITES} set.
+    *
+    * In the event of a partition merge, we define the preferred partition as the partition whom's coordinator is coordinating
+    * the current merge.
```
When you're resolving conflicts on demand, what is the preferred partition?
When you're merging a cluster with two degraded partitions, but one of the partitions contains all owners of given entry and the second one has the future coordinator, does it mean that you'll wipe out the data? (with PREFERRED_ALWAYS policy)

rvansa commented on core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java:
```diff
@@ -0,0 +1,416 @@
+               .stream(new ReplicaSpliterator(topology), false)
+               .filter(filterConsistentEntries());
+      } finally {
+         streamInProgress.compareAndSet(true, false);
```
You're returning the stream itself; no terminal operation was called so far on the stream, so it's still in progress, isn't it?

ryanemerson replied:
Good spot. Obviously this is wrong in any case, but this is a left over from when I was experimenting with a parallel spliterator. I have remove this statement to the `ReplicaSpliterator::tryAdvance`

rvansa commented on core/src/main/java/org/infinispan/conflict/ConflictManager.java:
```diff
@@ -0,0 +1,60 @@
+
+   /**
+    * @return a stream of Map<Address, CacheEntry>> for all conflicts detected throughout this cache.
+    * @throws org.infinispan.commons.CacheException if state transfer is initiated while checking for conflicts.
```
Who throws the `CacheException`, this method or some invocation on the stream?
And you don't mention the exception when the 'stream is in progress'

rvansa commented on core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java:
```diff
@@ -0,0 +1,416 @@
+
+   private void processConflict(Phaser phaser, Map<Address, CacheEntry<K, V>> conflictMap,
+                                                LocalizedCacheTopology topology, Set<Address> preferredPartition) {
+      stateTransferExecutor.execute(() -> {
```
nitpick: trivial statement + long lambda as a method doesn't look right; I would move the delegation to another executor to the forEach

rvansa commented on core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java:
```diff
@@ -0,0 +1,416 @@
+         if (preferredEntries.size() == 1) {
+            preferredEntry = conflictMap.remove(preferredEntries.get(0));
+         } else {
+            // If multiple conflicts exist in the preferred partition, then use primary replica from the preferred partition
```
This comment should go to ConflictManager javadoc

ryanemerson replied:
These details are already defined in `EntryMergePolicy` docs.

rvansa commented on core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java:
```diff
@@ -0,0 +1,416 @@
+         } else {
+            // If multiple conflicts exist in the preferred partition, then use primary replica from the preferred partition
+            // If not a merge, then also use primary as preferred entry
+            // Preferred is null if no entry exists in preferred partition
```
This is a bit inconsistent; if there is a single conflict in the preferred partition (e.g. because the entry on primaryReplica has been removed completely), you use that; if you have two other nodes (different from primary replica) with conflict, you take the value from primary replica which was not marked as a conflict before. Not sure if you can solve this without tombstones, though :-/ But at least document that.

rvansa replied:
I have noticed that you add a NullCacheEntry when you receive null, so this is probably okay.

rvansa commented on core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java:
```diff
@@ -0,0 +1,416 @@
+               if (entry.getValue() instanceof SuccessfulResponse) {
+                  SuccessfulResponse response = (SuccessfulResponse) entry.getValue();
+                  Object rspVal = response.getResponseValue();
+                  versionsMap.put(entry.getKey(), (InternalCacheValue<V>) rspVal);
```
`versionsMap` is not synchronized, you're accessing that from multiple threads (delivering response) concurrently

rvansa commented on core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java:
```diff
@@ -0,0 +1,416 @@
+            versionsMap.put(localAddress, icv);
+         }
+
+         ClusteredGetCommand cmd = commandsFactory.buildClusteredGetCommand(key, FlagBitSets.SKIP_OWNERSHIP_CHECK);
```
please use `cmd.setTopologyId(installedTopology.getTopologyId())` to mark which topology you use for sending

rvansa commented on core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java:
```diff
@@ -0,0 +1,416 @@
+                  Object rspVal = response.getResponseValue();
+                  versionsMap.put(entry.getKey(), (InternalCacheValue<V>) rspVal);
+               } else {
+                  completableFuture.completeExceptionally(new CacheException(String.format("Unable to retrieve key %s from %s: %s", key, entry.getKey(), entry.getValue())));
```
non-successful response isn't the end of the world, you can get `UnsureResponse` when the target does not own given entry anymore, or `CacheNotFoundResponse` when the node was suspected...

rvansa commented on core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java:
```diff
@@ -0,0 +1,416 @@
+                  iterator = segmentEntries.iterator();
+                  nextSegment++;
+               } catch (InterruptedException e) {
+                  Thread.currentThread().interrupt();
```
`staterReceiver.stop()`, too?

rvansa commented on core/src/main/java/org/infinispan/conflict/impl/StateReceiverImpl.java:
```diff
@@ -0,0 +1,238 @@
+
+   @DataRehashed
+   @SuppressWarnings("unused")
+   synchronized void onDataRehash(DataRehashedEvent dataRehashedEvent) {
```
is this package-private on purpose?

ryanemerson replied:
Yes. It is required by the `StateReceiverTest`

rvansa commented on core/src/main/java/org/infinispan/conflict/impl/StateReceiverImpl.java:
```diff
@@ -0,0 +1,238 @@
+      final List<Address> replicaHosts;
+      final Map<K, Map<Address, CacheEntry<K, V>>> keyReplicaMap = new HashMap<>();
+      final Map<Address, InboundTransferTask> transferTaskMap = new HashMap<>();
+      volatile CompletableFuture<List<Map<Address, CacheEntry<K, V>>>> future;
```
you're accessing that only in synchronized block, it does not have to be `volatile`

rvansa commented on core/src/main/java/org/infinispan/partitionhandling/PartitionHandling.java:
```diff
@@ -0,0 +1,11 @@
+ * @since 9.1
+ */
+public enum PartitionHandling {
+   DENY_READ_WRITES, // If the partition does not have all owners for a given segment, both reads and writes are denied for all keys in that segment.
```
Please use `/** */` style comment to document the behaviour in javadoc properly.

rvansa commented on core/src/main/java/org/infinispan/partitionhandling/impl/PartitionHandlingManager.java:
```diff
@@ -27,6 +27,14 @@
+    *
+    * @throws org.infinispan.partitionhandling.AvailabilityException if all owners of the keys are missing
+    */
+   void checkForMissingKeys(Collection<?> expected, Collection<?> actual);
```
Have you merely refactored the code or did you manage to understand it? Recently I got under the impression that the code has piled up over time and we should just throw `AvailabilityException` without the check at all.

ryanemerson replied:
Hmm looking back at it I'm not too sure, it was a long time ago I wrote this part of the code. Maybe it's worth me removing this refactor and changing the behaviour in a separate PR?

rvansa replied:
Checkout https://github.com/rvansa/infinispan/commit/11b4a436a299973c9401e33a33f3f828907139b3 please

rvansa replied:
Checkout https://github.com/infinispan/infinispan/pull/5175/files#diff-6427260c40c25c2ec242bc08e83b2b9c please

rvansa commented on core/src/main/java/org/infinispan/partitionhandling/impl/PreferAvailabilityStrategy.java:
```diff
@@ -20,13 +24,31 @@
+      if (topologyId != 0)
+         return topologyId;
+
+      String rt1 = t1.getCurrentCH().getRoutingTableAsString();
```
`getRoutingTableAsString` is only for debugging/logging!

rvansa commented on core/src/main/java/org/infinispan/topology/ClusterTopologyManagerImpl.java:
```diff
@@ -417,16 +418,19 @@ private ClusterCacheStatus initCacheStatusIfAbsent(String cacheName) {
-         if (cacheConfiguration != null && cacheConfiguration.clustering().partitionHandling().enabled()) {
+         Configuration cacheConfiguration = cacheManager.getCacheConfiguration(cacheName);
+         PartitionHandling  partitionHandling = cacheConfiguration != null ? cacheConfiguration.clustering().partitionHandling().whenSplit() : null;
+         boolean resolveConflictsOnMerge = partitionHandling == PartitionHandling.ALLOW_READ_WRITES && cacheConfiguration.clustering().partitionHandling().mergePolicy() != null;
```
You can likely get conflicts even without allowing writes: primary owner applies the write but backup (in the other partition) does not get them. The application will get an exception, but it does not have an easy way to schedule the resolution when it will be possible.

rvansa commented:
Check out 
https://github.com/rvansa/infinispan/commit/11b4a436a299973c9401e33a33f3f828907139b3 
please

On 06/15/2017 05:12 PM, Ryan Emerson wrote:
>
> *@ryanemerson* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In 
> core/src/main/java/org/infinispan/partitionhandling/impl/PartitionHandlingManager.java 
> <https://github.com/infinispan/infinispan/pull/4635#discussion_r122229103>:
>
> > @@ -27,6 +27,14 @@
>   
>      void checkBulkRead();
>   
> +   /**
> +    * Unlike in {@link #checkRead(Object)}, here we ignore the availability status and we only fail the operation if
> +    * _all_ owners have left the cluster.
> +    *
> +    * @throws org.infinispan.partitionhandling.AvailabilityException if all owners of the keys are missing
> +    */
> +   void checkForMissingKeys(Collection<?> expected, Collection<?> actual);
>
> Hmm looking back at it I'm not too sure, it was a long time ago I 
> wrote this part of the code. Maybe it's worth me removing this 
> refactor and changing the behaviour in a separate PR?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/infinispan/infinispan/pull/4635#discussion_r122229103>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ACEUPR3DCQB1BxdlFosjfz6UJo4jw72Eks5sEUnygaJpZM4KiLts>.
>


-- 
Radim Vansa <rvansa@redhat.com>
JBoss Performance Team

rvansa commented on core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java:
```diff
@@ -457,18 +477,19 @@ public void onTopologyUpdate(final CacheTopology cacheTopology, final boolean is
 
+         PartitionHandlingConfiguration phConfig = configuration.clustering().partitionHandling();
+         boolean wasMember = previousWriteCh != null && previousWriteCh.getMembers().contains(rpcManager.getAddress());
+         boolean deletePastMemberVals = wasMember && phConfig.whenSplit() != PartitionHandling.ALLOW_READ_WRITES && phConfig.mergePolicy() == null;
```
Can a merge policy be null? I thought it is set to preferred_always as default

ryanemerson replied:
Only if it is explicitly set to null in the configuration. This is so that the previous merge behaviour is still available.

rvansa replied:
OK, I've mistaken that for MergePolicy.NONE in parser. Shouldn't this be the default to keep the behaviour backwards compatible in minor release, then?

wburns commented:
There was a test failure in the last run: http://ci.infinispan.org/job/Infinispan/job/PR-4635/77/testReport/junit/org.infinispan.conflict.impl/OperationsDuringMergeConflictTest/testPartitionMergePolicy/

tristantarrant commented:
Merged to master. Good job !

tristantarrant commented:
I merged it because both CI and locally I got clean runs, but now I see intermittent failures on MergePolicyCustomTest>BaseMergePolicyTest.testPartitionMergePolicy:61->afterMerge:30->BaseMergePolicyTest.assertSameVersionAndNoConflicts:152 VersionMap: {MergePolicyCustomTest-NodeA-31787=ImmortalCacheValue {value=BEFORE SPLIT}, MergePolicyCustomTest-NodeC-45410=ImmortalCacheValue {value=BEFORE SPLIT}} expected:<Custom> but was:<BEFORE SPLIT>
